### ISSUE TITLE
Compilation error "no method names chars" E0599

### ISSUE THREAD

**Issue by hanysz** [Chicoryn/dream-go#33]
Trying to compile the source code from the zip of release 0.5.0, on Ubuntu 18.  Full log of errors copied below.  Any suggestions?  I'm completely new to the rust language: figuring out the difference between the "stable" and "nightly" compilers was already a big effort for me...

--------

DreamGo> tar zxf v0.5.0.tar.gz 
DreamGo> cd dream-go-0.5.0/
dream-go-0.5.0> rustup override add nightly
info: using existing install for 'nightly-x86_64-unknown-linux-gnu'
info: override toolchain for '/opt/DreamGo/dream-go-0.5.0' set to 'nightly-x86_64-unknown-linux-gnu'

  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.30.0-nightly (2ab3eba30 2018-09-14)

dream-go-0.5.0> cargo build
   Compiling void v1.0.2
   Compiling libc v0.2.34                                                                 
   Compiling bitflags v0.4.0                                                              
   Compiling lazy_static v1.0.0                                                           
   Compiling encode_unicode v0.1.3                                                        
   Compiling unicode-width v0.1.4                                                         
   Compiling utf8-ranges v1.0.0                                                           
   Compiling regex-syntax v0.4.1                                                          
   Compiling num-traits v0.1.41                                                           
   Compiling unreachable v1.0.0                                                           
   Compiling unreachable v0.1.1                                                           
   Compiling thread_local v0.3.5                                                          
   Compiling memchr v2.0.1                                                                
   Compiling num_cpus v1.7.0                                                              
   Compiling nix v0.5.1                                                                   
   Compiling time v0.1.38                                                                 
   Compiling rand v0.4.1                                                                  
   Compiling ordered-float v0.5.0                                                         
   Compiling aho-corasick v0.6.4                                                          
   Compiling threadpool v1.7.1                                                            
   Compiling rustyline v1.0.0                                                             
   Compiling regex v0.2.3                                                                 
   Compiling dream_go v0.5.0 (/opt/DreamGo/dream-go-0.5.0)                                
warning: unused import: `Read`                                                            
  --> src/nn/loader.rs:18:26                                                              
   |                                                                                      
18 | use std::io::{BufReader, Read};                                                      
   |                          ^^^^                                                        
   |                                                                                      
   = note: #[warn(unused_imports)] on by default                                          
                                                                                          
error[E0599]: no method named `chars` found for type `std::io::BufReader<std::fs::File>` in the current scope
  --> src/nn/loader.rs:76:45                                                              
   |                                                                                      
76 |         let mut iter = BufReader::new(file).chars().map(|ch| ch.unwrap());           
   |                                             ^^^^^                                    
                                                                                          
error: aborting due to previous error                                                     
                                                                                          
For more information about this error, try `rustc --explain E0599`.                       
error: Could not compile `dream_go`.                                                      

To learn more, run the command again with --verbose.

**Comment by hanysz** [comment]
Thanks!  Yes, I can build the head (after installing the libblosc-dev Ubuntu package), and it seems to play a decent game of go.  But it's playing very quickly, just a fraction of a second per move.  I'm driving it via gogui, which I think sends time_settings and time_left gtp commands, and it looks as though DreamGo is ignoring those commands (or else I'm doing something wrong--but I can persuade leela-zero to go slower, just not DreamGo).

**Comment by Chicoryn** [comment]
By default it performs 1,600 rollouts, which given modern hardware goes by very quickly. However it should listen to the GTP `time_settings`, `kgs-time_settings`, and `time_left` commands using the heuristics described in #29. There is some disagreement between how these commands should be implemented however, and I am pretty sure _Dream Go_ implements them differently from _Leela_.

To make _Dream Go_ play a five minute game you can use the following GTP command:

```
time_settings 300 0 1
```

Dream Go will assume the game is on average 257 moves, so with five minutes thinking time that is slightly more than 1 seconds per move on average. If you want it to play five seconds per move:

```
kgs-time_settings byoyomi 0 5 1
```

---

Good catch on the `libblosc-dev` dependency. I need to add that to the README.

**Comment by hanysz** [comment]
Thanks for the reply!  The details in #29 are pretty interesting.
OK, so time_settings seems to work, but time_left doesn't, at least not for me.
```
~> dream_go
time_settings 3600 0 1
= 

genmove w
  (pauses a few seconds as expected)
Nodes: 57482, Win: 61.5%, PV: d4 d3 c3 e3 e4 c2 f3 c4 b3 b2 b4 c5 b5 c6 f2 d1 e2 d2 b6 c7 b7
```

```
~> dream_go
time_left 300
? unknown command
```

**Comment by hanysz** [comment]
Sorry, my mistake, "unknown command" can either mean the command name is unknown, or else there is a syntax error. ```time_left b 300 1``` seems to be recognised.

I think DreamGo is doing the right thing here, but there is a problem with gogui.  If I launch gogui to play a game against DreamGo, and type ```time_settings 300 0 1``` into the gtp window, then it will play a 5-minute game.  But if I use gogui-twogtp to start an engine-engine match, I can tell from the logs that it's sending the same command, but DreamGo plays at a faster pace.  I've checked that gogui sends time_settings at the start of the game, and doesn't send time_left.  Somehow this problem doesn't seem to affect other engines, only DreamGo.

**Comment by Chicoryn** [comment]
Hmm, interesting. Could you attach the logs that you get from gogui-twogtp so that I can maybe try to re-produce and fix the issue?

**Comment by Chicoryn** [comment]
I took a look at the code for gogui-twogtp and pushed some changes that I think will help. So if you want to re-build the latest from master that might help.

In summary the problem was that we were not handling all of the "edge" cases for the `time_settings` GTP command correctly. This has now been corrected so that:

- `time_settings x 0 0` should mean _x_ seconds of absolute time.
- `time_settings x y 0` should mean no limit if _y > 0_.
- Otherwise Canadian time
